lib/api/unittest/CDataFrameAnalyzerTest.cc

```diff
@@ -745,6 +763,80 @@ void CDataFrameAnalyzerTest::testRoundTripDocHashes() {
     }
 }
 
+void CDataFrameAnalyzerTest::testCategoricalFields() {
+
+    std::stringstream output;
+    auto outputWriterFactory = [&output]() {
+        return std::make_unique<core::CJsonOutputStreamWrapper>(output);
+    };
+
+    {
+        api::CDataFrameAnalyzer analyzer{regressionSpec(1000, 100000, {"x1", "x2"}),
+                                         outputWriterFactory};
+
+        TStrVec x[]{{"x11", "x12", "x13", "x14", "x15"},
+                    {"x21", "x22", "x23", "x24", "x25", "x26", "x27"}};
+
+        for (std::size_t i = 0; i < 10; ++i) {
+            analyzer.handleRecord({"x1", "x2", "x3", "x4", "x5", ".", "."},
+                                  {x[0][i % x[0].size()], x[1][i % x[1].size()],
+                                   std::to_string(i), std::to_string(i),
+                                   std::to_string(i), std::to_string(i), ""});
+        }
+        analyzer.receivedAllRows();
+
+        bool passed{true};
+
+        const core::CDataFrame& frame{analyzer.dataFrame()};
+        frame.readRows(1, [&](TRowItr beginRows, TRowItr endRows) {
+            std::size_t i{0};
+            for (auto row = beginRows; row != endRows; ++row, ++i) {
+                core::CFloatStorage expected[]{static_cast<double>(i % x[0].size()),
+                                               static_cast<double>(i % x[1].size())};
+                passed &= (expected[0] == (*row)[0]);
+                passed &= (expected[1] == (*row)[1]);
+            }
+        });
+
+        CPPUNIT_ASSERT(passed);
+    }
+
+    LOG_DEBUG(<< "Test overflow");
+    {
+        std::size_t rows{api::CDataFrameAnalyzer::MAX_CATEGORICAL_CARDINALITY + 3};
+
+        api::CDataFrameAnalyzer analyzer{regressionSpec(rows, 1000000000, {"x1"}),
+                                         outputWriterFactory};
+
+        TStrVec fieldNames{"x1", "x2", "x3", "x4", "x5", ".", "."};
+        TStrVec fieldValues{"", "", "", "", "", "", ""};
+        for (std::size_t i = 0; i < rows; ++i) {
+            std::fill_n(fieldValues.begin(), 6, std::to_string(i));
+            analyzer.handleRecord(fieldNames, fieldValues);
+        }
+        analyzer.receivedAllRows();
+
+        bool passed{true};
+        std::size_t i{0};
+
+        const core::CDataFrame& frame{analyzer.dataFrame()};
+        frame.readRows(1, [&](TRowItr beginRows, TRowItr endRows) {
+            for (auto row = beginRows; row != endRows; ++row, ++i) {
+                if (i < api::CDataFrameAnalyzer::MAX_CATEGORICAL_CARDINALITY) {
+                    core::CFloatStorage expected{static_cast<double>(i)};
+                    passed &= (expected == (*row)[0]);
+                } else {
+                    core::CFloatStorage expected{static_cast<double>(
+                        api::CDataFrameAnalyzer::MAX_CATEGORICAL_CARDINALITY)};
+                    passed &= (expected == (*row)[0]);
+                }
+            }
+        });
+
+        CPPUNIT_ASSERT(passed);
```

Comment:
If this assertion ever fails it won't give any indication of what went wrong.  Therefore it might be worth logging an error at the point(s) where `expected != (*row)[0]`.

Same in the method above too.